File: packages/ui/src/schemas/loans.ts
import { z } from 'zod';

export const LoanProductSchema = z.object({
  id: z.string(),
  name: z.string(),
  type: z.string(),
  term_years: z.number(),
  rate_type: z.string(),
  min_down_payment_pct: z.number(),
  min_credit_score: z.number().nullable(),
  max_dti_ratio: z.number().nullable(),
  max_loan_amount: z.number().nullable(),
  description: z.string().nullable(),
  eligibility_requirements: z.array(z.string()),
  features: z.array(z.string()),
  is_active: z.boolean(),
});

export const LoanProductListSchema = z.object({
  items: z.array(LoanProductSchema),
  total: z.number(),
});

export const EligibilityCheckResponseSchema = z.object({
  eligible: z.boolean(),
  estimated_rate: z.string().nullable(),
  estimated_monthly_payment: z.number().nullable(),
  max_loan_amount: z.number().nullable(),
  warnings: z.array(z.string()),
  suggestions: z.array(z.string()),
});

export type LoanProduct = z.infer<typeof LoanProductSchema>;
export type LoanProductList = z.infer<typeof LoanProductListSchema>;
export type EligibilityCheckResponse = z.infer<typeof EligibilityCheckResponseSchema>;

export interface EligibilityCheckRequest {
  annual_income: number;
  monthly_debts: number;
  credit_score_range: string;
  down_payment_amount: number;
  property_value: number;
  citizenship_status: string;
}

File: packages/api/src/schemas/loans.py
"""
Loan product Pydantic schemas.
"""

from pydantic import BaseModel, Field


class LoanProductResponse(BaseModel):
    """Loan product response."""

    id: str
    name: str
    type: str
    term_years: int
    rate_type: str
    min_down_payment_pct: float
    min_credit_score: int | None = None
    max_dti_ratio: float | None = None
    max_loan_amount: float | None = None
    description: str | None = None
    eligibility_requirements: list[str] = []
    features: list[str] = []
    is_active: bool = True


class LoanProductListResponse(BaseModel):
    """Paginated list of loan products."""

    items: list[LoanProductResponse]
    total: int


class EligibilityCheckRequest(BaseModel):
    """Pre-qualification eligibility check request."""

    annual_income: float = Field(gt=0)
    monthly_debts: float = Field(ge=0)
    credit_score_range: str = Field(
        description="Self-reported credit score range, e.g. '620-679'"
    )
    down_payment_amount: float = Field(ge=0)
    property_value: float = Field(gt=0)
    citizenship_status: str = Field(
        description="citizen, permanent_resident, visa_holder, non_resident"
    )


class EligibilityCheckResponse(BaseModel):
    """Pre-qualification eligibility check result."""

    eligible: bool
    estimated_rate: str | None = None
    estimated_monthly_payment: float | None = None
    max_loan_amount: float | None = None
    warnings: list[str] = []
    suggestions: list[str] = []
